### PR TITLE
Add NFT proposal

### PR DESCRIPTION
**Pull request by arjanvaneersel**
When opening a pull request to submit a new AEX, please use the suggested template: [https://github.com/aeternity/AEXs/blob/master/aex-X.md](https://github.com/aeternity/AEXs/blob/master/aex-X.md)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "vscode-corda.isCordaProject": false
+}
```
Comment (mradkov) [inline .vscode/settings.json]: Can you exclude this file from the PR?

File: AEXS/aex-nft.md [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+## NFTReceiver
+
+contract interface NFTReceiver = 
+    entrypoint on_nft_received : (address, address, int, option(string)) => unit
+
+### on_nft_received\(\)
+
+Deals with receiving NFT tokens on behalf of a smart contract. Contracts receiving NFT tokens should implement this interface. Mint and transfer transactions will invoke the `on_nft_received` function.
+
+```sofia
```
Comment (mradkov) [inline AEXS/aex-nft.md]:
```suggestion
```sophia
```

File: AEXS/aex-nft.md [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+| :--- | :--- |
+| account| address |
+| value | int |
+
+# Receiver contract interface
+
+## NFTReceiver
+
+contract interface NFTReceiver = 
+    entrypoint on_nft_received : (address, address, int, option(string)) => unit
```
Comment (mradkov) [inline AEXS/aex-nft.md]:
```suggestion
```sophia
contract interface NFTReceiver = 
    entrypoint on_nft_received : (address, address, int, option(string)) => unit
```
```

File: AEXS/aex-nft.md [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+| owner | address |
+| approved | address |
+
+| return | type |
+| :--- | :--- |
+| approved | bool |
+     
+## Events
+
+```
```
Comment (mradkov) [inline AEXS/aex-nft.md]:
```suggestion
```sophia
```

File: AEXS/aex-nft.md [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+| return | type |
+| :--- | :--- |
+| meta_info | meta_info |
+
+### meta_data\(\)
+
+Returns meta data associated with a token. The function is a part of the basic interface, because metadata can be set in the constructor, as well as by implementing the Mintable extention.
+
+```entrypoint metadata(id int): option(metadata)```
```
Comment (mradkov) [inline AEXS/aex-nft.md]:
```suggestion
```sophia
entrypoint metadata(id int): option(metadata)
```
```

File: AEXS/aex-nft.md [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+## Motivation
+
+The following standard describes standard interfaces for non-fungible tokens. The proposal contains a primary interface and secondary interfaces for optional functionality that not everyone might need. 
+
+# Basic NFT
+
+## Interface
+
+```
```
Comment (mradkov) [inline AEXS/aex-nft.md]:
```suggestion
```sophia
```

**Review by mradkov — CHANGES_REQUESTED** [review]
Let's keep the formatting consistent.

File: .vscode/settings.json (hunk 1 above)
Comment (arjanvaneersel) [reply .vscode/settings.json]: Done that

File: AEXS/aex-nft.md [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+contract interface NFTReceiver = 
+    entrypoint on_nft_received : (address, address, int, option(string)) => unit
+
+### on_nft_received\(\)
+
+Deals with receiving NFT tokens on behalf of a smart contract. Contracts receiving NFT tokens should implement this interface. Mint and transfer transactions will invoke the `on_nft_received` function.
+
+```sophia
+entrypoint on_nft_received : (address, address, int, option(string)) => unit
+```
```
Comment (marc0olo) [inline AEXS/aex-nft.md:407]: I think something went wrong with the old suggestion. I also prefer `NFTs` over `NFT tokens` as the "T" already stands for token.

```suggestion
Deals with receiving NFTs on behalf of a smart contract. A smart contract MUST implement this interface to be able to receive NFTs. Mint and transfer transactions will invoke the `on_nft_received` function.

```sophia
contract interface NFTReceiver = 
    entrypoint on_nft_received : (address, address, int, option(string)) => unit
```
```

File: AEXS/aex-nft.md [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+
+## NFTReceiver
+
+```sophia
+contract interface NFTReceiver = 
+    entrypoint on_nft_received : (address, address, int, option(string)) => unit
+
+### on_nft_received\(\)
+
+Deals with receiving NFT tokens on behalf of a smart contract. Contracts receiving NFT tokens should implement this interface. Mint and transfer transactions will invoke the `on_nft_received` function.
```
Comment (marc0olo) [inline AEXS/aex-nft.md]: I also prefer `NFTs` over `NFT tokens` as the "T" already stands for token.

```suggestion
Deals with receiving NFTs on behalf of a contract. Contracts MUST implement this interface to be able to receive NFTs. Mint and transfer transactions will invoke the `on_nft_received` function.
```

File: AEXS/aex-nft.md [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+| meta_info | meta_info |
+
+### meta_data\(\)
+
+Returns meta data associated with a token. The function is a part of the basic interface, because metadata can be set in the constructor, as well as by implementing the Mintable extention.
+
+```sophia
+entrypoint metadata(id int): option(metadata)
+
+| parameter | type |
```
Comment (marc0olo) [inline AEXS/aex-nft.md:99]: formatting was broken

```suggestion
```sophia
entrypoint metadata(id int): option(metadata)
```

| parameter | type |
```

**Comment by marc0olo** [comment]
@arjanvaneersel The suggestions somehow cannot deal with this. But in some sophia codeblocks the ``` on the bottom are missing. This breaks the formatting, There are multiple occurences.

Example:
image

File: AEXS/aex-nft.md [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+
+## Abstract
+
+A standard implementation of non-fungible tokens for the Aeternity ecosystem. The design goal of the primary interface is to be as compatible with ERC-721 as possible, so that anyone who can work with ERC-721 can work with this interface. However, where Sophia offers a better way, performance and efficiency should prevail over compatibility.
+
+There is no support for unsafe transactions. Therefore all transactions are ought to be safe.
+
+## Motivation
+
+The following standard describes standard interfaces for non-fungible tokens. The proposal contains a primary interface and secondary interfaces for optional functionality that not everyone might need.
```
Comment (omar-saadoun) [inline AEXS/aex-nft.md]: Small text edit : 
```suggestion
The following standard describes standard interfaces for non-fungible tokens. The proposal contains a primary and secondary interfaces for optional functionality that not everyone might need. 
```

**Review by omar-saadoun — COMMENTED** [review]
Small editions to text description

**Comment by thepiwo** [comment]
for me it looks good with the comments from @omar-saadoun and @marc0olo adressed

**Comment by marc0olo** [comment]
@arjanvaneersel I saw you accepted some changes. there is still the formatting issue I mentioned in my comment.

can we agree on AEX-141 @thepiwo, @omar-saadoun? it's the number of the issue and got most votes on discord.

if so you can also change the filename and the `aexX` and `aex-X` placeholders in the standard definition.

after that I think we are ready to go and can merge it finally =)

**Comment by arjanvaneersel** [comment]
@marc0olo There's no change request with your comments, at least I don't see a button to accept your comments.

**Comment by marc0olo** [comment]
Hi @arjanvaneersel, I am talking about the following comment: https://github.com/aeternity/AEXs/pull/143#issuecomment-995363684

If you address this and change the placeholders and file accordingly to fit the `AEX-141` we can merge =)

**Comment by marc0olo** [comment]
@arjanvaneersel when will you be able to make the last changes? you can also provide me the rights to adjust this on your branch if you want.

how should we proceed with the implementation? I will probably find somebody of the team to tackle this if that is ok for you. we need to move forward a bit faster on that topic

File: AEXS/aex-141.md [hunk 11]
```diff
@@ -1,7 +1,7 @@
 # AEX Non-Fungible Token Standard
```
Comment (marc0olo) [inline AEXS/aex-141.md]:
```suggestion
# AEX-141: Non-Fungible Token Standard
```

File: AEXS/aex-141.md [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+## Methods
+
+### aexX_extensions\(\)
+
+This function **returns** a hardcoded list of all implemented extensions on the deployed contract.
+X is to be replaced with the approriate number allocated to this AEX proposal. At this stage the number is unknown.
+
+```sophia
+entrypoint aex9_extensions() : list(string)
+```
```
Comment (marc0olo) [inline AEXS/aex-141.md]:
```suggestion
### aex141_extensions\(\)

This function **returns** a hardcoded list of all implemented extensions on the deployed contract.

```sophia
entrypoint aex141_extensions() : list(string)
```
```

File: AEXS/aex-141.md [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+        , symbol: string 
+        , base_url: option(string)
+        , type : metadata_type)}
+
+    datatype event 
+        = Transfer(indexed address, indexed address, indexed int)
+        | Approval(indexed address, indexed address, indexed int, bool)
+        | ApprovalForAll(indexed address, indexed address, bool)
+
+    entrypoint aexX_extensions() : list(string)
```
Comment (marc0olo) [inline AEXS/aex-141.md]:
```suggestion
    entrypoint aex141_extensions() : list(string)
```

File: AEXS/aex-141.md [hunk 14]
```diff
@@ -0,0 +1,415 @@
+# AEX Non-Fungible Token Standard
+
+```
+AEX: 141
+Title: Non-Fungible Token Standard
+Author: Arjan van Eersel <arjan@appiness.solutions> (@zkvonsnarkenstein)
+License: ISC
+Discussions-To: <URL>
+Status: Draft
```
Comment (marc0olo) [inline AEXS/aex-141.md]:
```suggestion
Discussions-To: https://forum.aeternity.com/t/aeternity-nft-token-standard/9781
Status: Review
```

**Review by marc0olo — CHANGES_REQUESTED** [review]
few more changes and we should finally be ready to merge =) ... thanks Arjan!

**Review by marc0olo — APPROVED** [review]
we did it, finally! 🥳  thanks @arjanvaneersel! :-)

**Pull request merged**